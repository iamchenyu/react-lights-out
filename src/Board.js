import React, { useState } from "react";
import Cell from "./Cell";
import "./Board.css";

/** Game board of Lights out.
 *
 * Properties:
 *
 * - nrows: number of rows of board
 * - ncols: number of cols of board
 * - chanceLightStartsOn: float, chance any cell is lit at start of game
 *
 * State:
 *
 * - board: array-of-arrays of true/false
 *
 *    For this board:
 *       .  .  .
 *       O  O  .     (where . is off, and O is on)
 *       .  .  .
 *
 *    This would be: [[f, f, f], [t, t, f], [f, f, f]]
 *
 *  This should render an HTML table of individual <Cell /> components.
 *
 *  This doesn't handle any clicks --- clicks are on individual cells
 *
 **/

function Board({ nrows = 5, ncols = 5, chanceLightStartsOn = 0.6 }) {
  const [board, setBoard] = useState(createBoard());

  /** create a board nrows high/ncols wide, each cell randomly lit or unlit */
  function createBoard() {
    let initialBoard = [];
    // TODO: create array-of-arrays of true/false values
    for (let x = 0; x < nrows; x++) {
      let row = [];
      for (let y = 0; y < ncols; y++) {
        // Old version
        // row.push(Math.random() < chanceLightStartsOn ? true : false);
        // Refactor version
        row.push(Math.random() < chanceLightStartsOn);
      }
      initialBoard.push(row);
    }
    return initialBoard;
  }

  function hasWon() {
    // TODO: check the board in state to determine whether the player has won.

    // Old version
    // let allCells = [];
    // board.map((row) => row.map((cell) => allCells.push(cell)));
    // return allCells.find((el) => el);

    // New version
    return board.every((row) => row.every((cell) => !cell));
  }

  function flipCellsAround(coord) {
    setBoard((board) => {
      const [x, y] = coord.split("-").map(Number);

      const flipCell = (x, y, arr) => {
        // if this coord is actually on board, flip it

        if (x >= 0 && x < nrows && y >= 0 && y < ncols) {
          arr[x][y] = !arr[x][y];
        }
      };

      // TODO: Make a (deep) copy of the oldBoard
      const boardCopy = [...board];

      // TODO: in the copy, flip this cell and the cells around it
      flipCell(x, y, boardCopy);
      flipCell(x - 1, y, boardCopy);
      flipCell(x + 1, y, boardCopy);
      flipCell(x, y - 1, boardCopy);
      flipCell(x, y + 1, boardCopy);

      // TODO: return the copy
      return boardCopy;
    });
  }

  // if the game is won, just show a winning msg & render nothing else

  // TODO
  if (hasWon()) {
    alert("You have won!");
  }
  // make table board

  // TODO
  return (
    <table className="Board">
      <tbody>
        {board.map((row, idxRow) => {
          return (
            <tr key={idxRow}>
              {row.map((cell, idxCol) => {
                return (
                  <Cell
                    key={idxCol}
                    isLit={cell}
                    flipCellsAroundMe={() =>
                      flipCellsAround(`${idxRow}-${idxCol}`)
                    }
                  />
                );
              })}
            </tr>
          );
        })}
      </tbody>
    </table>
  );
}

export default Board;
