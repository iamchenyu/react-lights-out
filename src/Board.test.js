import React from "react";
import { render, fireEvent } from "@testing-library/react";
import Board from "./Board";

beforeEach(() => {
  jest.spyOn(global.Math, "random").mockReturnValue(0.5);
});

afterEach(() => {
  jest.spyOn(global.Math, "random").mockRestore();
});

it("renders without crashing", () => {
  render(<Board />);
});

it("matches the snapshot", () => {
  const { asFragment } = render(<Board />);
  expect(asFragment()).toMatchSnapshot();
});

// New version
it("handles the click", () => {
  const { queryAllByRole } = render(<Board />);
  const cells = queryAllByRole("button");

  // check the status of each cell
  cells.forEach((cell) => expect(cell).toHaveClass("Cell Cell-lit"));

  // click on a random cell
  fireEvent.click(cells[6]);

  // check if only the cell and the surrouding cells are switched status
  let index = [1, 5, 6, 7, 11];
  cells.forEach((cell, idx) => {
    if (index.includes(idx)) {
      expect(cell).not.toHaveClass("Cell-lit");
    } else {
      expect(cell).toHaveClass("Cell-lit");
    }
  });
});

it("display the message properly", () => {
  const alertMock = jest.spyOn(window, "alert").mockImplementation();
  const { queryAllByRole } = render(<Board nrows={1} ncols={3} />);
  const cells = queryAllByRole("button");

  // no alert when start the game
  expect(alertMock).toHaveBeenCalledTimes(0);

  // click on the middle cell
  fireEvent.click(cells[1]);

  // display alert
  expect(alertMock).toHaveBeenCalledTimes(1);
});

// Old version
it("handles the click", () => {
  const alertMock = jest.spyOn(window, "alert").mockImplementation();
  const { getByTestId } = render(<Board nrows={1} ncols={1} />);
  const cell = getByTestId("cell");
  expect(cell.className).toBe("Cell Cell-lit");

  // check the cell flips
  fireEvent.click(cell);
  expect(cell.className).toBe("Cell ");

  // check the alert message comes up
  expect(alertMock).toHaveBeenCalledTimes(1);
});
