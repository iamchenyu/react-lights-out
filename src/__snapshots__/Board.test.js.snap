// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches the snapshot 1`] = `
<DocumentFragment>
  <table
    class="Board"
  >
    <tbody>
      <tr>
        <td
          class="Cell Cell-lit"
          data-testid="cell"
          role="button"
        />
        <td
          class="Cell Cell-lit"
          data-testid="cell"
          role="button"
        />
        <td
          class="Cell Cell-lit"
          data-testid="cell"
          role="button"
        />
        <td
          class="Cell Cell-lit"
          data-testid="cell"
          role="button"
        />
        <td
          class="Cell Cell-lit"
          data-testid="cell"
          role="button"
        />
      </tr>
      <tr>
        <td
          class="Cell Cell-lit"
          data-testid="cell"
          role="button"
        />
        <td
          class="Cell Cell-lit"
          data-testid="cell"
          role="button"
        />
        <td
          class="Cell Cell-lit"
          data-testid="cell"
          role="button"
        />
        <td
          class="Cell Cell-lit"
          data-testid="cell"
          role="button"
        />
        <td
          class="Cell Cell-lit"
          data-testid="cell"
          role="button"
        />
      </tr>
      <tr>
        <td
          class="Cell Cell-lit"
          data-testid="cell"
          role="button"
        />
        <td
          class="Cell Cell-lit"
          data-testid="cell"
          role="button"
        />
        <td
          class="Cell Cell-lit"
          data-testid="cell"
          role="button"
        />
        <td
          class="Cell Cell-lit"
          data-testid="cell"
          role="button"
        />
        <td
          class="Cell Cell-lit"
          data-testid="cell"
          role="button"
        />
      </tr>
      <tr>
        <td
          class="Cell Cell-lit"
          data-testid="cell"
          role="button"
        />
        <td
          class="Cell Cell-lit"
          data-testid="cell"
          role="button"
        />
        <td
          class="Cell Cell-lit"
          data-testid="cell"
          role="button"
        />
        <td
          class="Cell Cell-lit"
          data-testid="cell"
          role="button"
        />
        <td
          class="Cell Cell-lit"
          data-testid="cell"
          role="button"
        />
      </tr>
      <tr>
        <td
          class="Cell Cell-lit"
          data-testid="cell"
          role="button"
        />
        <td
          class="Cell Cell-lit"
          data-testid="cell"
          role="button"
        />
        <td
          class="Cell Cell-lit"
          data-testid="cell"
          role="button"
        />
        <td
          class="Cell Cell-lit"
          data-testid="cell"
          role="button"
        />
        <td
          class="Cell Cell-lit"
          data-testid="cell"
          role="button"
        />
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
